---
title: "DESeq2 Tutorial"
author: "Zedias Chikwambi"
date: "`r Sys.Date()`"
output: html_document
---

Useful resources 

- **RNAseq Workflow Vignette** - http://master.bioconductor.org/packages/release/workflows/vignettes/rnaseqGene/inst/doc/rnaseqGene.html
- **Deseq2 Vignette** - http://www.bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html

# install packages 

- If i didn't already have these installed in my environment - I would delete the # on lines 19-23 and 26-39 and run the code 

```{r}
# from cran 
# install.packages('tidyverse')
# install.packages("RColorBrewer")
# install.packages('pheatmap')
# if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")

# from bioconductor
# BiocManager::install("ashr")
# BiocManager::install("apeglm")
# BiocManager::install("DESeq2")
# BiocManager::install("org.Mm.eg.db")
```

# load packages

- what do these packages do? Add comments to remind you what they do 
```{r}
library("pheatmap")   
library("RColorBrewer")
library(tidyverse)     
library(DESeq2)      
library(ashr)          
library(apeglm)      
library(org.Mm.eg.db)
```

# 1 Prepping our input data to make deseq2dataset object

## Import counts table

- We made this (or a very similar version) using featurecounts in week1 on the cluster
- raw data -> fastqc -> hisat2 -> featurecounts 

```{r}
# Import counts data to variable called featurecounts_table
# The file has a comment line we need to ignore at top of file 
featurecounts_table <- read_tsv("data/allsamples_counts.tsv.gz", comment = "#")
head(featurecounts_table)

# lets save the gene info (Chr,Start,End,Strand, Length) columns as a seperate dataframe (df) incase we need it later
# call this table gene_info and set row names as Geneid
gene_info <- featurecounts_table %>%
                column_to_rownames("Geneid") %>%                     
                dplyr::select(Chr, Start, End, Strand, Length)
head(gene_info)

# create a counts table without the gene info (Chr,Start,End,Strand, Length) columns
# to get a clean counts table where:
# rows = genename, columns = samples
counts_table <- featurecounts_table %>%
                  column_to_rownames("Geneid") %>%
                  dplyr::select(-Chr, -Start, -End, -Strand, -Length) %>%
                  as.matrix()

head(counts_table)
dim(counts_table)
# check the dimensions of this counts_table table using dim() -> should have 39,179 genes and 8 cols 


```

# Right now lets import and set up our table of sample information

```{r}
sample_info <- read_tsv("~/OBDS_training/differential_expression/data/PRJEB18572_sample_info.tsv")
head(sample_info)

# What class is the cell_type currently? 



# We want to change it to a factor with 2 levels (the hypothesis we will test)
# Which should be our reference level? 


# what other columns might we want to include in our design that we want to change to factors? 

class(sample_info$cell_type)
sample_info$cell_type<-factor(sample_info$cell_type, levels=c("CD4", "CD8"))
levels(sample_info$cell_type)
```
```{r}
sample_info$genotype<-as.factor(sample_info$genotype)
sample_info$replicate<-as.factor(sample_info$replicate)
head(sample_info)
```


# Check the sample names are the same and in the same order in the sample_info and counts_table 

```{r}
# Column names of the count table and row names of the sample info table must be in the same order
# Check this 

# make sample_info row names from sample_id 

# remove .bam from colnames 
colnames(counts_table)

```
```{r}
sample_info<-column_to_rownames(sample_info, var = "sample_id")
head(sample_info)
```
```{r}
#head(counts_table)
colnames(counts_table)<-colnames(counts_table) %>% 
  str_replace(".bam","")

head(counts_table)
```

# Generate DESeqDataSet object

```{r}
# look up function to use from slides, find help and create an object called dds
#dds <- DESeqDataSet()

dds<-DESeqDataSetFromMatrix(
  countData=counts_table,
  colData=sample_info,
  design=~genotype + cell_type,
  tidy = FALSE,
  ignoreRank = FALSE
) 
dds

```

## Investigate the dds object 

```{r}
# what does it look like? - e.g. use str(dds)

str(dds)
# how to access counts
counts(dds)

# access colData i.e the sample info table
colData(dds)

# access rowData
rowData(dds)

# access the design

design(dds)
# does it have any rowRanges()?

rowRanges(dds)
# what are its rownames and colnames?
rownames(dds)
colnames(dds)

# what attributes does it have?
attributes(dds)

```

## How to subset a DESeqDataSet Object 

```{r}
# practice subsetting dds object
# select first 3 columns
dds_subset<-dds[, 1:3]
dds_subset
```

# Filter out unexpressed genes 

- It's good practice to filter out genes with 0 counts across all samples at the start 
- This reduces the size of the object and makes computations quicker without affecting the results as these genes are filtered out anyway in the analysis 
- Note that more strict filtering to increase power is automatically applied via independent filtering on the mean of normalized counts within the results function.


```{r}
# filter out genes that have 0 counts across all samples and save as an object called `dds_filtered`
dds_filtered<-dds[rowSums(assays(dds)$count) >0, ]
dds_filtered

# check dims of dds_filtered  
  dim(dds_filtered)
  
```

# Perform exploratory analysis on dds_filtered object 

1. transform dds_filtered object using `rlog()` functions
    - name your new variables `rlog_transformed_dds` 
    - we want to set `blind = TRUE` as because we are doing a QC step we do not want the function to take into account the experimental design, we want it to be an unbiased look at how the samples cluster
    - check the class of the rlog_transformed_dds objects 
    - you can look at the tranformed counts table using `assay()` 

NOTE we only use this tranformed data for QC not for downstream analysis or differential expression itself!!


```{r}
#transform data using `rlog()` functions
rlog_transformed_dds <- rlog(dds_filtered, blind = T)


# Notes on `blind` parameter: 
  # blind = TRUE is default
  # blind = TRUE is not the appropriate choice if you expect many or majority of genes to have large differences in counts explainable by   experimental design
  # Mike Love recommends blind = FALSE for visualisation, clustering and machine learning but blind = TRUE for QC
  # Experimental design is only used for calculating the gene-wise dispersion estimates in order to fit a trend line through the dispersions over the mean

  
#check the class of these variable objects 

  class(rlog_transformed_dds)
#get and look at the tranformed counts table using `assay()`

assay(rlog_transformed_dds)
```

# Optional: If you have time do the tranformation again but using the vst() function 

- As a guide 
    - rlog - more robust than VST when size factors vary widely (10x)
    - VST - less sensitive to high count outliers than rlog, quicker - recommended most often 

```{r}
# get vst transformed data using vst() 
vst_transformed_data <- 

```


# Now plot heatmap of sample-sample correlations using rlog transformed count data and cor()


```{r}
# get correlation matrix of the rlog_transformed counts matrix to get correlations between samples 
cor_matrix<-cor(assay(rlog_transformed_dds), method = c("spearman"))

# use pheatmap to plot the correlation matrix 
pheatmap(cor_matrix)
```

# try ploting a heatmap of sample distances 

```{r}
# get sample distances -> Annotated what this code is doing?
sampleDists <- dist(t(assay(rlog_transformed_dds))) 
sample_distances_matrix <- as.matrix(sampleDists) 

pretty_blue_colors <- colorRampPalette(rev(brewer.pal(9, "Blues")) )(255)

pheatmap(sample_distances_matrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists, 
         color = pretty_blue_colors)
```

What do you notice here!? what does it tell us about our data? 

Answer: 



## You can subset a dds object and interact with it like you would a DataFrame

```{r}
# remove the two suspect columns using [,] notation - remove suspicious samples 
rlog_transformed_dds_subset <- rlog_transformed_dds[,1:6]

# lets replot the correlation heatmap with the subsetted data - copy code and adjust from above
cor_matrix_sub<-cor(assay(rlog_transformed_dds_subset), method = c("spearman"))

# use pheatmap to plot the correlation matrix 
pheatmap(cor_matrix_sub)

```

What do you notice now about the data and the clustering?


# Now lets use the plotPCA function from DESeq2 to plot PCA of rlog_transformed_data_subset

```{r, fig.asp=1}
# plot pca of rlog_transformed_dds using plotPCA()
plotPCA(rlog_transformed_dds_subset, intgroup="cell_type") + theme(aspect.ratio = 1)

# what does the intgroup parameter do? have a go at changing it...


# how many variable genes is it using for pca? 
#ntop=500

# plot the pca for all genes 
plotPCA(rlog_transformed_dds_subset, intgroup="cell_type", 
        ntop=nrow(rlog_transformed_dds_subset)) + theme(aspect.ratio = 1)

```


What does it tell us about our sample?

Answer: 


------------- This is the end of our Exploratory data analysis -----------------

We now go back to our original dds object with the raw counts and can update it with anything we need to from the exploratory analysis 
    e.g. - update design with variables in sample info table we want to regress out 
         - remove samples (poor quaility or those with name switches etc)
         
# Revisiting our inital data 

What do we need to do to our data??? 

## Update dds_filtered object if needed - call it dds_filtered_updated

```{r}

# do we need to remove/edit samples? 
dds_filtered_updated<-dds_filtered[, 1:6]

# do we want to update the design? Is this nessecary in this case?
# e.g. design(dds) <- ~replicate + condition
design(dds_filtered_updated) <- ~replicate + cell_type
```


# Now we can start our actual DE!!!! yay 

# We will run the analysis step by step to see whats going on 

## 1. Estimate size factors using estimateSizeFactors()

```{r}
# Calculate the size factors for each sample using estimateSizeFactors()
dds_filtered_updated<-estimateSizeFactors(dds_filtered_updated)
# Access the size factors from the dds_filtered_updated object
sizeFactors(dds_filtered_updated)
# Do you see any systematic differences in size factors between groups?

```

# Optional: Plot size factors as barchart using ggplot
```{r}
```

## Get counts table normalised using size factors

```{r}
# get counts table normalised using size factors - call it normalized_counts

normalized_counts<-counts(dds_filtered_updated, normalized=T)
# check the class of the output
class(normalized_counts)
```

# 3. Estimate dispersions for each gene

```{r}
# Obtain dispersion estimates for each gene using estimateDispersions() 
dds_filtered_updated<-estimateDispersions(dds_filtered_updated)
# Plot the per-gene dispersion estimates (DESeq2 has a helper function for this)
plotDispEsts(dds_filtered_updated)
```


## 4. Perform Wald test

```{r}
# Tests for significance of coefficients in a negative binomial GLM, using previously calculated size factors and dispersion estimates and WaldTest
dds_filtered_updated<-nbinomWaldTest(dds_filtered_updated)
dds_filtered_updated
```


# So we have the actual steps that we've gone through to get to DE in one place generate the dds object again and use the DESeq() function to perform steps 1-3 in one go

```{r}
# generate dds object 

# filter rows with 0 counts 

# remove unwanted samples 

# make DESeqDataSet (dds) object 

# run DESeq() 

```

# Get the results from your dds object - call it `results_obj`

```{r}
# What do the results() and summary() do, what arguments can you specify? 


```

# Optional - if we have time: Access the coefficients of the Negative Binomnial GLM – see vingette("DESeq2")
```{r}

```

# Plot normalised counts for ENSMUSG00000041779
 
 checkout the 'plotCounts()` function
 
```{r}
# Plot normalised counts for gene ENSMUSG00000041779

```

# Access the results table for the comparison between CD8 and CD4 T cells . Youu can also try to CD4 vs CD8 (i.e. switch the reference) - you could also try this for replicate 1 vs replicate 3 as a control 

checkout `resultsNames()`  & `results()`

```{r}

```

Table should contain log2 fold changes and p-values for all genes

# Plot a histogram of the raw and BH-adjusted p-values using ggplot – do they look as expected? for CD8 vs CD4

```{r}

```


# Generate an MA plot of the log2 FC values for all genes using DESeq2 plotMA()

```{r}


```

# Shrink the log2 FC values using the normal, apeglm and ashr methods -> see the vignette – which shall we choose? 

```{r}

# call your results table  called shrunkenLFC_dds_results_obj

```


# Generate a results table (one shrinkage method) containing gene symbols and (if time) entrez ids 

```{r}

# convert shrunkenLFC_dds_results_obj to a dataframe and call it shrunkenLFC_dds_results_df

# Use the org.Mm.eg.db package to get gene symbols for the ensembl gene ids (hint:you did this on bioconductor day) 

# How many Ensembl IDs are not assigned an gene symbol and how many gene symbols are duplicated?



keytypes(org.Mm.eg.db)      #to remind us of what keytypes are in this particular db

# use mapIds() to get 1:1 mapping for SYMBOL, ENTREZ and MGI

# add mapId vectors to  shrunkenLFC_dds_results_df
```


# Export all the results to CSV

```{r}

# first change covert ensembl_ids to a column 
shrunkenLFC_dds_results_df <- results_shrunken_LFC_apeglm_filtered_df %>% 
						rownames_to_column("ensembl_gene_ids")

# Save results to file 
write_csv(shrunkenLFC_dds_results_df, "CD8_vs_CD4_DE_results_lfcshrinkage_allgenes.csv")
```


Well done! That is the basics of exploratory data analysis for bulk RNAseq and differential gene expression - these statistical methods underly the foundations for single cell differential expression analysis, although scRNA-seq has a few extra paramterisations and complications to take into account. 

We will now progress to pathway/geneset analysis to get some biological meaning from our DE gene list....


Output the sessionInfo() so know what packages we used
```{r}
sessionInfo()

```

# Bonus Exercise 

Generate a heatmap of normalised counts for the top 20 (by shrunken FC) differentially-expressed genes across all samples– label samples by condition and genes by mgi symbol
```{r}
# get the results table for 20 genes with the highest log2 fold changes from shrunken lfc results object  
#top_20_DE_genes_df <- 


  
  
# get the ensembl ids of these genes 
# top_20_DE_gene_ids = 

# get the gene symbols for these genes 
# top_20_DE_gene_symbols =

# get normalised counts matrix for the top 20 DE genes 
# norm_counts_for_top20_de_genes <- 
# set mgi symbols as row names 

# make heatmap of normalised counts using pheatmap()

```

Plot a volcano plot:
Highlight significantly differentially expressed genes (p adj < 0.05, log2FC > 2) in red
Add labels to highlight the location of some interesting genes

```{r}
# volcano plot has fold change on x-axis and -log10(padj) on y axis 
# create a new column for -log10(padj)
# create a new column to label genes with TRUE/FALSE if log2foldchange > 2 & padj < 0.05 
#df_for_volcano_plot <- 



# plot using ggplot - has fold change on x-axis and -log10(padj) on y axis 
#volcano plot <- 
```