---
title: Week4-day5 :integration template"
author: "Devika Agarwal/updataed by Carla Cohen"
date: "17/10/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE) #print code and results
```


```{r}
library(Seurat)
library(SeuratData)
library(SeuratWrappers)
library(patchwork)
library(tidyverse)
library(batchelor)
library(harmony)
library(cowplot)
library(lisi)
library(MAST)
#install.packages("devtools")
#devtools::install_github("immunogenomics/lisi")
#remotes::install_github('satijalab/seurat-wrappers')

options(future.globals.maxSize = 4000 * 1024^2)    # Maximum allowed total size (in bytes) of global variables identified
```

# Exercise 1 - joint analysis of two PBMC datasets

Here we will look at how to PBMC datasets compare without any integration. 

## 1. Import the two RDS files into R

These RDS object have already been put through the QC and filtering steps covered on Seurat day 1.


```{r}
pbmcv2 <- readRDS("~/OBDS_training/integration/pbmcv2_filtered.rds")
pbmcv3 <- readRDS("~/OBDS_training/integration/pbmcv3_filtered.rds")
str(pbmcv2)
pbmcv2@ meta.data$nCount_RNA
```


Plot these metadata covariates nCount_RNA/nFeature_RNA/percent.mt to see how they range for both the datasets 

use `VlnPlot()` from Seurat

```{r}

p1 <- VlnPlot(pbmcv2, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"))
p2 <- VlnPlot(pbmcv3, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"))
plot_grid(p1, p2, ncol = 1, labels = c("p1", "p2"))
```



## 2. Create a merged Seurat object containing both sets of data (no batch correction)

Vignette: https://satijalab.org/seurat/articles/merge_vignette.html

```{r}
seurat_merged <-  merge(x = pbmcv2,y = pbmcv3,
                       add.cell.ids = c("v2", "v3"))

```

Check the dims of the merged object in comparison to the separate objects

What is the total gene number in the merged object compared to each single object?

```{r}
dim(seurat_merged)
dim(pbmcv2)
dim(pbmcv3)
str(seurat_merged)
head(seurat_merged[[]])
```

## 3.1 Perform SCTransform normalisation

Do you want to regress any meta.data variables ?

```{r, warning=FALSE}
seurat_merged <- SCTransform(seurat_merged,
                             assay = "RNA",
                             vars.to.regress = "percent.mt")

```

Plot to see how the nCount_RNA looks before and after SCT normalisation for the merged dataset?

Did SCT do a good job in normalizing the sequencing depth across the two datasets?


```{r}
p3 <- VlnPlot(seurat_merged, features = c("nCount_RNA", "nCount_SCT"), same.y.lims = T)
p3


```


What should the Default Assay be?

```{r}
DefaultAssay(seurat_merged)
```

Plot the Variable features plot for data with top 10 variable genes labelled

```{r, fig.width=8}
plot1 <- VariableFeaturePlot(seurat_merged, selection.method = "sct")
plot1
top10 <- VariableFeatures(seurat_merged)[1:10] # select the top 10 genes
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE) # label the points corresponding to the top ten genes.
plot2 #plot the scatter plot showing the labeled top ten genes.
```

## 3.2 Run PCA and visualize 

Can you see a batch effect between pbmc v2 and v3?


```{r}
seurat_merged <- RunPCA(seurat_merged)
```

#Look at which genes are driving your PCs.

```{r, fig.width=8}
VizDimLoadings(seurat_merged, d= 1:2)
```


```{r}
DimPlot(seurat_merged)

```

Choose optimal  number of PCs for UMAP

```{r}
ElbowPlot(seurat_merged) # By default the ndims is = 50
```

## 3.3 Run the Seurat UMAP function and visualize

Is there a batch effect? 


```{r}
seurat_merged <- RunUMAP(seurat_merged, dims = 1:20)
DimPlot(seurat_merged, group.by = "orig.ident")
```

### Optional: Do cells cluster by celltypes?

Plot some pbmc celltype markers to see how they look on the UMAP

```{r}
FeaturePlot(seurat_merged, 
            features = c("CD3E", "CD4", "CD8A", "MS4A1", "CD14", "FCGR3A","LYZ","GNLY"))
# Is it not possible to screen and only return genes uniquely expressed in particular cell types.
FeaturePlot(seurat_merged, 
            features = top10)
```


## Exercise 2: Seurat integration with SCTransform

Check out the Vignette: https://satijalab.org/seurat/articles/integration_introduction.html#performing-integration-on-datasets-normalized-with-sctransform-1 

The Seurat v2 integration method is slightly different depending on whether you perform log normalistaion or SCTransformation on your seurat objects. Here we are using the SCTransform method (the vignette covers both).

### 1. Set up

Import the RDS files 

We will read in the objects again to start the integration workflow from scratch

```{r}
pbmcv2 <- readRDS("~/OBDS_training/integration/pbmcv2_filtered.rds")
pbmcv3 <- readRDS("~/OBDS_training/integration/pbmcv3_filtered.rds")
```

Generate a named list of the two objects to run workflow

```{r}
pbmc.list <- list("v2" = pbmcv2, "v3" = pbmcv3)
pbmc.list
```

Apply SCTransform across the list

```{r}
#this applies the function to each object in the list independently. 
#this is different to above where we ran the SCTransform on the merged data, so it was performed on all the data at once.
pbmc.list <- lapply(pbmc.list, SCTransform, variable.features.n=3000, vars.to.regress = "percent.mt") 
```

Check dims for both objects in list after SCTransform
```{r}
dim(pbmc.list[[1]]) 
dim(pbmc.list[[2]])
```

Make sure Default assay in both objects is SCT
```{r}
DefaultAssay(pbmc.list[[1]])
DefaultAssay(pbmc.list[[2]])
lapply(pbmc.list, DefaultAssay)
```

Choose the features to use when integrating multiple datasets. 

This function ranks features by the number of datasets they are deemed variable in, breaking ties by the median variable feature rank across datasets. It returns the top scoring features by this ranking.
```{r}
features <- SelectIntegrationFeatures(pbmc.list,
                                      nfeatures = 3000)
```

Prepare an object list normalized with sctransform for integration (run `PrepSCTIntegration()` )

This function subsets the scale.data slot to only contain residuals for anchor.features for efficiency in downstream processing
It is specific to the SCTransform workflow and makes sure you have the SCTransform residuals for the remaining genes

```{r}
pbmc.list <- PrepSCTIntegration(object.list = pbmc.list, 
                                anchor.features = features)

```

> After this step, the features present in the scale.data slots of the two objects are the same
> anchor.features are used in the anchor finding process


### Optional exercise

How similar are these features to the variable features that we found for the pbmcv2 and v3 individually?
- extract variable features for each dataset individually
- compare these with "features"; which has the variable features most common for both datasets

```{r}
v2_features <- VariableFeatures(pbmc.list[[1]])
v3_features <- VariableFeatures(pbmc.list[[2]])

intersect(features, v2_features) %>% length() #2386 variable features in common
intersect(features, v3_features) %>% length() #2023 variable features in common

```


## 2. Perform integration

Find a set of anchors between a list of Seurat objects. These anchors can later be used to integrate the objects.

```{r}
pbmc.anchors <- FindIntegrationAnchors(pbmc.list,
                                       anchor.features = features,
                                       normalization.method = "SCT",
                                       reduction = "cca", #default
                                       dims = 1:20) #same as merge so that we can compare outputs
```

Integrate the two pbmc objects

```{r}
pbmc.combined.sct <- IntegrateData(pbmc.anchors,
                                   dims = 1:20,
                                   normalization.method = "SCT")
```

What is the Default assay now?
Check to see if the Seurat object has changed in the Assay slot at all

```{r}
DefaultAssay(pbmc.combined.sct)
```

## 3. Analysis of integrated data

Run  PCA on Integrated data

```{r}
pbmc.combined.sct <- RunPCA(pbmc.combined.sct, npcs = 50)
```

How many PCs should we use going forward?
```{r}
ElbowPlot(pbmc.combined.sct, ndims = 50)

```

Visualise PCA, how does this compare to the merged data?

```{r}
DimPlot(pbmc.combined.sct)
DimPlot(seurat_merged, reduction = "pca")
```



Run UMAP 
```{r}
pbmc.combined.sct <- RunUMAP(pbmc.combined.sct,
                             dims = 1:15)
```


```{r}
DefaultAssay(pbmc.combined.sct)
```

Find neighbors and clusters for the integrated data

Run `FindNeighbors()` and `FindClusters()` 
Try using multiple resolutions at the same time for the clustering e.g. 0.5 and 0.8
What other clustering algorithms could be used here? Which one is used by default?


```{r}
pbmc.combined.sct <- FindNeighbors(pbmc.combined.sct,
                                   dims = 1:15)

# Could use Graphs (pbmc.combined.sct) to check that this step has been performed

pbmc.combined.sct <- FindClusters(pbmc.combined.sct,
                                  resolution = c(0.5, 0.8)
                                  )
```


How many clusters have been defined at each resolution?
```{r}
#look at the metadata
pbmc.combined.sct[["integrated_snn_res.0.5"]] %>% unique()  # 9 clusters
pbmc.combined.sct[["integrated_snn_res.0.8"]] %>% unique()  # 11 clusters

#Alternatively:

table(pbmc.combined.sct@meta.data$integrated_snn_res.0.5)
table(pbmc.combined.sct@meta.data$integrated_snn_res.0.8)

table(pbmc.combined.sct@meta.data$integrated_snn_res.0.5,pbmc.combined.sct@meta.data$integrated_snn_res.0.8)
#could possibly plot this in a ven-diagram - ### To look into this!!

```

Visualse UMAP plot for batch effect and clusters

check the Default Assay to be sure

```{r}
DefaultAssay(pbmc.combined.sct)

DimPlot(pbmc.combined.sct, reduction = "umap",
        group.by = c("orig.ident", "integrated_snn_res.0.5", "integrated_snn_res.0.8") )

```
```{r}
DefaultAssay(pbmc.combined.sct) <-"SCT" #using "SCT"
SCT<-FeaturePlot(object = pbmc.combined.sct, 
            features = c("S100A9", "S100A8", "LYZ", "IGKC","GNLY","IGHM", "NKG7", "IGLC2"))
```


Plot some canonical PBMC cluster markers e.g. S100A9, S100A8, LYZ, IGKC, GNLY, IGHM, NKG7, IGLC2

Which assay should we plot them on? (try both Integrated and SCT, is there a difference?)
First plot on the Integrated assay

```{r, fig.width=8}
DefaultAssay(pbmc.combined.sct) <- "integrated" #using "integrated"
integrated<-FeaturePlot(object = pbmc.combined.sct, 
            features = c("S100A9", "S100A8", "LYZ", "IGKC","GNLY","IGHM", "NKG7", "IGLC2"))

```


```{r}
cowplot::plot_grid(SCT, integrated, labels = c("SCT", "integrated"))
DefaultAssay(pbmc.combined.sct) <- "SCT" #set to SCT assay
```


Set default assay to "SCT" then do the same plot
Set default assay to "SCT" then do the same plot
> We can see that when the integrated data is plotted there is a higher background than when we plot the SCT. 
> So better to visualise on the SCT plots. 
> This is because of the background smoothing that has been done to batch correct.

## 4. Identify markers

We want to identify markers that distinguish each cluster. 
Look up `FindMarkers` and `FindAllMarkers`, which one should we use here?

> FindAllMarkers is looking for the best genes that distinguish each cluster by performing sequential pairwise comparisons between all the clusters
> Whereas FindMarkers is used to compare once cluster to all the others, and you specify which cluster is your baseline cluster
> here we will use FindALLMarkers

Why are we changing the DefaultAssay?
Do we need to renormalise RNA Assay?

> We can't run FindMarkers on the integrated SCT model because the genes are no longer independent of each other. 
> Therefore it does not meet the assumptions that are required for DE. 

> NB if using SCTransform v2 there is an additional step ` PrepSCTFindMarkers` and then FndAllMarkers can be run on the SCT assay
> Here we will run it on the RNA assay as usual. No further normalisation steps are required

Use the "LR" test

```{r}
DefaultAssay(pbmc.combined.sct) <- "RNA"

# specify which clustering we are using 
Idents(pbmc.combined.sct) <- pbmc.combined.sct$integrated_snn_res.0.5

pbmc.combined.sct.markers.LR <- FindAllMarkers(pbmc.combined.sct, test.use = "LR", 
                                  latent.vars = "orig.ident",
                                  only.pos = TRUE) # the interest is in normalising for the effect of both variables and returning genes up-regulated

# could also add only only.pos = TRUE
```

> latent.vars is things you want to regress out when it does the DE for each gene
> it cannot be set for Wilcoxon which is the default test
> various thresholds can be changed e.g. min.pct and logfc.threshold
> pros and cons of the various statistical tests are discussed here: https://www.nature.com/articles/nmeth.4612

Filter the top 5 markers for each cluster
use `dply::group_by` and `slice_max()`
How do you define "top"?

```{r}

pbmc.combined.sct.markers.top5 <- pbmc.combined.sct.markers.LR %>% 
    group_by(cluster) %>% 
    slice_max(order_by = avg_log2FC, n = 5)

pbmc.combined.sct.markers.top5

```


plot Heatmap or Dotplot of markers

try functions `Seurat::DoHeatmap` or `Seurat::Dotplot`

```{r}

DefaultAssay(pbmc.combined.sct) <- "SCT"
DoHeatmap(pbmc.combined.sct,
          features = pbmc.combined.sct.markers.top5$gene)


```
Dotplot

```{r}
DotPlot(pbmc.combined.sct,
        features = unique(pbmc.combined.sct.markers.top5$gene))+ #include unique as some of the genes are duplicated
    RotatedAxis() #rotate the x axis text by 45 degrees so easier to read


```


How well do you think the clustering has worked?
Are you able to identify each cluster?
Would you re-evaluate any of the clusters?


> We can see that clusters 1&2 are quite similar, so are 4, 5, 6 so perhaps we could review the clustering in this case. 


## Optional 

Spend some time experimenting with different statistical tests in the FindAllMarkers command
Try some different parameters
Sort by log2fc vs padj
How does this affect the plots?
Use your results to label the identity of each cluster and plot on a UMAP


FindAllMarkers using MAST test

```{r}
DefaultAssay(pbmc.combined.sct) <- "RNA"
pbmc.combined.sct.markers.MAST <- FindAllMarkers(pbmc.combined.sct, method = "MAST", 
                                    latent.vars = c("nCount_RNA", "orig.ident"),
                                    min.pct = 0.25) #each gene must be present in 25% of cells
                                     
```

Select top 5 markers ordered by padj

```{r}


pbmc.combined.sct.markers.top5.MAST.log2fc <- pbmc.combined.sct.markers.MAST %>% 
    group_by(cluster) %>% 
    slice_max(order_by = avg_log2FC, n = 5)



pbmc.combined.sct.markers.top5.MAST.padj <- pbmc.combined.sct.markers.MAST %>% 
    group_by(cluster) %>% 
    slice_min(order_by = p_val_adj , n = 5) #NB use slice_min here for padj

pbmc.combined.sct.markers.top5.MAST.log2fc
pbmc.combined.sct.markers.top5.MAST.padj
```

Plot heatmaps

```{r}

DefaultAssay(pbmc.combined.sct) <- "SCT"
DoHeatmap(pbmc.combined.sct,
          features = pbmc.combined.sct.markers.top5.MAST.log2fc$gene)
DoHeatmap(pbmc.combined.sct,
          features = pbmc.combined.sct.markers.top5.MAST.padj$gene)

```
> Heat map is less clear if we use padj as the cut off rather than log2FC


Plot dotplot

```{r}
DotPlot(pbmc.combined.sct,
        features = unique(pbmc.combined.sct.markers.top5.MAST.log2fc$gene))+ #include unique as some of the genes are duplicated
    RotatedAxis() #rotate the x axis text by 45 degrees so easier to read
```

Label the cluster identities

First choose which resolution you want to work from (here I am using 0.5 i.e. 9 clusters)
Refer to the UMAP plot and canonical markers plots generated above
```{r}
Idents(pbmc.combined.sct) <- "integrated_snn_res.0.5"

```

Look for some markers described here:
https://satijalab.org/seurat/articles/pbmc3k_tutorial.html#assigning-cell-type-identity-to-clusters-1

```{r}
DimPlot(pbmc.combined.sct, reduction = "umap", group.by = "integrated_snn_res.0.5")
DefaultAssay(pbmc.combined.sct) <- "SCT"
FeaturePlot(object = pbmc.combined.sct, 
            features = c("S100A4", "IL7R", "CCR7", "CD8A", "CD14", "LYZ", "FCGR3A","MS4A7", "GNLY","NKG7", "MS4A1", "FCER1A", "CST3", "PPBP"))
```

Rename the identities and plot

```{r}
pbmc.combined.sct <- RenameIdents(pbmc.combined.sct, `0` = "CD14 Mono", `1` = "CD4 Memory T", `2` = "CD4 Naive T",
    `3` = "B.1", `4` = "CD8 T.1", `5` = "NK", `6` = "CD8 T.2", `7` = "B.2", `8` = "CD8 T.3")


DimPlot(pbmc.combined.sct, label = TRUE)
```
> We would have to look for additional markers to resolve the B and CD8 T cell subsets
> we also migh want to understnad why some Naive CD4 T cells are clustering with the monocytes.



#Bonus Exercise: Run Harmony based integration

You can run Harmony within your Seurat workflow. You'll only need to make two changes to your code.

IMPORTANT DIFFERENCE: In the Seurat integration tutorial, you need to define a Seurat object for each dataset. 
With Harmony integration, create only one Seurat object with all cells.
Before running Harmony, make a Seurat object and following the standard pipeline through PCA.

After that Run Harmony with the `RunHarmony()` function
In downstream analyses, use the Harmony embeddings instead of PCA.
For example, run Harmony and then UMAP in two lines.

Vignette: http://htmlpreview.github.io/?https://github.com/immunogenomics/harmony/blob/master/docs/SeuratV3.html


Create a merged object, run SCTransform and PCA
```{r, warning = FALSE}
seurat_merged <- merge(x = pbmc.list[[1]], y = pbmc.list[[2]])

seurat_merged <- SCTransform(seurat_merged, vars.to.regress = "percent.mt")
seurat_merged <- RunPCA(seurat_merged,)

DimPlot(seurat_merged, group.by = "orig.ident")
```

Run Harmony
```{r}

# specific error on cbrg cluster is solved by running this code
# see https://github.com/immunogenomics/harmony/issues/173
harmony.onehot.new <- function (x) 
{
  data.frame(x) %>% tibble::rowid_to_column("row_id") %>% dplyr::mutate(dummy = 1) %>% 
    tidyr::spread(x, .data$dummy, fill = 0) %>% dplyr::select(-row_id) %>% 
    as.matrix
}
environment(harmony.onehot.new) <- asNamespace('harmony')
assignInNamespace("onehot", harmony.onehot.new, ns = "harmony")

```

Run Harmony and include the optimisation graph

```{r}
#better to change the name of the object here so that we can go back and make comparisons with the seurat_merged

seurat_merged_harmony <- seurat_merged %>% 
    RunHarmony(group.by.vars = "orig.ident", plot_convergence = TRUE, dims.use = 1:30, assay.use = "SCT")
```

Run UMAP, FindNeighbours and FindClusters, then visualise some key genes. 

```{r}
seurat_merged_harmony <- RunUMAP(seurat_merged_harmony, reduction = "harmony", dims = 1:30)
seurat_merged_harmony <- FindNeighbors(seurat_merged_harmony, reduction = "harmony", dims = 1:30)
seurat_merged_harmony <- FindClusters(seurat_merged_harmony, resolution = c(0.5, 0.8))

FeaturePlot(seurat_merged_harmony, 
            features = c("CD3E", "CD4", "CD8A", "MS4A1", "CD14", "FCGR3A",
                         "NCAM1", "KLRB1"))
```



> CD4 and CD8 T cells are nicely separated
> Small CD16+ monocyte cluster with cells from both datasets

How do the Harmony and Seurat integration methods compare?

```{r}
harm_DimPlot(seurat_merged_harmony, group.by = "orig.ident")
DimPlot(seurat_merged_harmony, group.by = "seurat_clusters")
DimPlot(pbmc.combined.sct, group.by = "orig.ident")
DimPlot(pbmc.combined.sct, group.by = "seurat_clusters")
```
```{r}

p_1 <- DimPlot(seurat_merged, reduction = "pca")
p_2 <- DimPlot(seurat_merged_harmony, reduction = "pca", group.by = "orig.ident")
p_3 <- DimPlot(pbmc.combined.sct, reduction = "pca", group.by = "orig.ident")

cowplot::plot_grid(p_1, p_2, p_3, labels = c("seurate", "harmony", "merged"))
```

```{r}

p_4 <- DimPlot(seurat_merged, reduction = "umap")

Reductions(seurat_merged)
p_5 <- DimPlot(seurat_merged_harmony, reduction = "umap", group.by = "orig.ident")
p_6 <- DimPlot(pbmc.combined.sct, reduction = "umap", group.by = "orig.ident")

cowplot::plot_grid(p_4, p_5, p_6, labels = c("merged", "harmony", "integrated"))
```





Compute LISI : https://github.com/immunogenomics/LISI 

```{r}
# Compute integration LISI and cell-type LISI
# iLISI - effective number of datasets in a neighbourhood - want this to equal number of batches
# cLISI - want this to equal 1 - number of cell types in neighbourhood i.e. different cell types should form distinct clusters
meta_data_cells <- data.frame(technology = seurat_merged_harmony@meta.data$orig.ident)
lisi_res <- lisi::compute_lisi(Embeddings(seurat_merged_harmony, reduction = "harmony"), 
                               meta_data_cells, 
                               "technology")
```



```{r}

# Should you use clusters determined from joint embedding for clisi?

cat("LISI values - iLISI mean and range, cLISI mean and range")
mean(lisi_res$technology)   # 1.70 (2 would be perfect if equal number of cells in two datasets)
range(lisi_res$technology)

# For cLISI, need to separately annotate cell types in the two individual datasets
```

